src/foam/dao/QueryCachingDAODecorator.js

```diff
@@ -0,0 +1,156 @@
+
+  constants: [
+    {
+      name: 'CLEAR_CACHE',
+      value: 'CLEAR_CACHE'
+    }
+  ],
+
+  properties: [
+    {
+      // The cache for local storage and fast access
+      name: 'cache',
+      factory: function() { return {}; }
+    },
+    {
+      // Number of elements in the DAO backing this query cache
+      // Allows limit to what we cache so we do not have to always call DAO if limit is not provided
+      name: 'daoCount_',
+      value: 0
+    }
+  ],
+
+  methods: [
+    // Put invalidates cache and is forwarded to the source.
+    function put_(x, o) {
+      //* Can caches be kept up to date efficiently? ****
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      return this.delegate.put_(x, o);
+    },
+
+    function select_(x, sink, skip, limit, order, predicate) {
+      // Only handle query caching for ArraySink and Projection
+      if ( ! (foam.dao.ArraySink.isInstance(sink) || foam.mlang.sink.Projection.isInstance(sink)) ) {
+        return this.SUPER(x, sink, skip, limit, order, predicate);
+      }
+
+      var self = this;
+      var key  = [sink, order, predicate].toString();
+
+      return new Promise(function(resolve, reject) {
+        console.log('******** QUERYCACHE: key: ' + key + ' in cache: ' +  ( self.cache[key] ? 'true' : 'false' ) + ' daoCount_: ' + self.daoCount_);
+
+        // Validate we have a fresh dao count
+        self.refreshDaoCount_(self).then(function() {
+
+          var requestStartIdx = typeof skip !== 'undefined' ? skip : 0;
+          var requestEndIdx = typeof limit !== 'undefined' && skip + limit < self.daoCount_ ? skip + limit : self.daoCount_;
+
+          // Ensure we have cache for request
+          self.fillCache_(self, key, requestStartIdx, requestEndIdx, x, sink, order, predicate).then(function() {
+
+            // Return data from cache
+            for (let idx = requestStartIdx; idx < requestEndIdx; idx++) {
+              sink.put(self.cache[key][idx]);
+            }
+
+            resolve(sink);
+          });
+        });
+      });
+    },
+
+    // Remove invalidates cache and is forwarded to the source.
+    function remove_(x, o) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      return this.delegate.remove_(x, o);
+    },
+
+    // RemoveAll invalidates cache and is forwarded to the source.
+    function removeAll_(x, skip, limit, order, predicate) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      this.delegate.removeAll_(x, skip, limit, order, predicate);
+    },
+
+    function cmd_(x, obj) {
+      if ( obj == this.CLEAR_CACHE ) {
+        this.cache = {};
+        this.clearProperty('daoCount_');
+      } else {
+        this.SUPER(x, obj);
+      }
+    },
+
+    function fillCache_(self, key, requestStartIdx, requestEndIdx, x, sink, order, predicate) {
+      // Pre-check cache for elements
+      let startIdx = -1;
```

Comment:
RE: my comment below:

I actually think you might be able to use `this` within the method instead of having to pass in self. I think this will refer to the model given its described in this helper method.

Reply:
i removed passing self as parameter and set local variable inside function instead where needed